core/types/external_tx.go

```diff
@@ -26,16 +26,45 @@ type ExternalTx struct {
 	// the origin chain indeed confirmed emission of that ETX.
 }
 
-// PendingEtxs are ETXs which have been emitted from a block, but are not yet
-// referencable through the block manifest. Once a coincident block is found,
-// the dominant nodes will collect all pending ETXs from each block listed in
-// the manifest, and mark them referencable so they may be included at the
-// destination chain.
+// PendingEtxs are ETXs which have been emitted in a subordinate block. The
+// block is not valid in dominant chains, but dominant chains relay the pending
+// ETXs to other chains in the network to facilitate ETX forward propagation.
+//
+// A dominant chain does not have the state to check correctness or acceptability
+// of these ETXs in the subordinate chains, but it does need to know that these
+// ETXs are valid against a block header which came from a subordinate chain.
+// For this reason, we indlude a header from the subordinate chain.
 type PendingEtxs struct {
 	Header *Header        `json:"header" gencodec:"required"`
+	// Etxs array contains ETXs from the chain which produced this block, and a
+	// subordinate rollup of ETXs for that chain's subordinate (if it has one).
+	// Etxs[originCtx] = external transactions in origin CTX
+	// (optional) Etxs[originCtx+1] = rollup of ETXs emitted by originCtx+1
 	Etxs   []Transactions `json:"etxs"   gencodec:"required"`
 }
 
+func (p *PendingEtxs) IsValid(hasher TrieHasher) bool {
+	nodeCtx := common.NodeLocation.Context()
+	if p == nil || p.Header == nil || p.Etxs == nil {
+		return false
+	}
+	if len(p.Etxs) < common.HierarchyDepth {
+		return false
+	}
+	// pending ETXs must have originated from our subordinate context.
+	singletonCtx := nodeCtx + 1
+	rollupCtx := singletonCtx + 1
+	// singletonCtx must exist and must match hash
+	if singletonCtx >= len(p.Etxs) || DeriveSha(p.Etxs[singletonCtx], hasher) != p.Header.EtxHash(singletonCtx) {
```

Comment:
even this length check.

Reply:
This length check makes sure `singletonCtx` wouldn't exceed the length of our array, because we are about to index into the array.